### PR TITLE
[exporter/signalfx] Update code to use HTTPClientSettings

### PR DESCRIPTION
**Pull request by manang-splunk**
**Description:**  Updated the signalfx exporter(metrics and logs) to use HTTPClientSettings to create HttpClient.

**Testing:**  Updated the test cases as per new changes

### CODE REVIEW

File: exporter/signalfxexporter/exporter.go [hunk 1]
```diff
@@ -56,11 +56,18 @@ func (sme *signalfMetadataExporter) ConsumeMetadata(metadata []*metadata.Metadat
...
+type metricExporter struct {
 	pushMetricsData    func(ctx context.Context, md pdata.Metrics) (droppedTimeSeries int, err error)
 	pushMetadata       func(metadata []*metadata.MetadataUpdate) error
-	pushLogsData       func(ctx context.Context, ld pdata.Logs) (droppedLogRecords int, err error)
 	hostMetadataSyncer *hostmetadata.Syncer
+	config             *Config
+	logger             *zap.Logger
+}
+
+type logExporter struct {
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/exporter.go:67]: why are this division and associated updates necessary?

Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go:67]: The end goal here is to use [HTTPClientSettings](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/confighttp.go#L36) to make 'httpClient' instead of creating it in the same code. In order to do that, we need a host of type component.Host in [httpClient.ToClient()](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/confighttp.go#L99
) method which can only be available in a [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method AFAIK. So, The approach we decided to use was to create a Start method and call newSignalFxExporter or newEventExporter in the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method based on the context of the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method (from [createMetricsExporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/factory.go#L99) or [createLogsExporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/factory.go#L179
)).
If we use the same structure for Logs as well as Metrics, we have two potential solutions:
1. Call both method ([newSignalFxExporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L95) or [newEventExporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L184
)) in the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) regardless of the context(createMetricsExporter or createLogsExporter).
2. Identify the context of the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method and call the respective function from the Start method.

In the first solution, it does not seem optimal as we would be calling an unnecessary function (newSignalFxExporter or newEventExporter) from the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method. For ex., If a user configures only one pipeline (metrics) still newEventExporter will be called which is unnecessary.

In the second solution, we need to identify the context of the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method. Using the existing signalfxExporter structure, it seems unlikely to identify the context of the [Start](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b96dfffb063e33f771425a54f65afbf9741dd482/exporter/signalfxexporter/exporter.go#L90
) method (createMetricsExporter or createLogsExporter).

Hence, we used different structures for the Logs and Metrics pipeline. So that at runtime, we will have different Start methods for different pipelines and we can get context.

Comment (rmfitzpatrick) [reply exporter/signalfxexporter/exporter.go:67]: I don't believe this is the correct approach, and the ask is to embed the `confighttp.HTTPClientSettings` into the exporter config for usage in client instantiation and not create a new instance directly. Briefly walking through this none of the changes you're making appear necessary.  We aren't using custom authenticators so passing in an empty extension map to `ToClient(map[config.ComponentID]component.Extension{}`)  should be ok afaict.

For example:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/zipkinexporter/config.go#L31
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/jaegerthrifthttpexporter/config.go#L25

It's important that these settings are backward compatible for field renames so legacy fields should be used to set the client setting field values if applicable.

File: exporter/signalfxexporter/exporter_test.go [hunk 2]
```diff
@@ -51,65 +52,82 @@ func TestNew(t *testing.T) {
...
 		config         *Config
+		me             *metricExporter
 		wantErr        bool
 		wantErrMessage string
 	}{
 		{
-			name:           "nil config fails",
+			name: "nil config fails",
+			me: &metricExporter{
+				logger: zap.NewNop(),
```
Comment (MovieStoreGuy) [inline exporter/signalfxexporter/exporter_test.go:62]: I would strongly recommend the use of `zaptest.NewLogger` instead of `Nop` in the event that something goes wrong in the test, it should be obvious what is the issue.

**Review by MovieStoreGuy — COMMENTED** [review]
I worry that this is some unrelated changes going in and would make it harder to debug tests in future.

File: exporter/signalfxexporter/exporter.go (hunk 1 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go]: Added `confighttp.HTTPClientSettings` into exporter config and changed `MaxConnections` to pointer for backward compatibility.

File: exporter/signalfxexporter/exporter.go [hunk 3]
```diff
@@ -146,6 +143,29 @@ func newSignalFxExporter(
 	}, nil
 }
 
+func getHTTPClient(config *Config) (*http.Client, error) {
+	defaultMaxConns := 100
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/exporter.go:147]: I believe we want the defaults to be loaded in the default config (https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/add40e0381a591accefacab8c14c39daa278b40b/exporter/signalfxexporter/factory.go#L54). The actual default values should probably be defined as const.

File: exporter/signalfxexporter/exporter.go [hunk 4]
```diff
@@ -146,6 +143,29 @@ func newSignalFxExporter(
...
+	if config.MaxIdleConnsPerHost == nil {
+		if config.MaxConnections != nil {
+			config.MaxIdleConnsPerHost = config.MaxConnections
+		} else {
+			config.MaxIdleConnsPerHost = &defaultMaxConns
+		}
+	}
+	// To avoid overriding client headers to request header storing access_token
+	// when access_token_passthrough is true.
+	config.HTTPClientSettings.Headers = map[string]string{}
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/exporter.go:164]: Appears we should respect these values and override any existing: https://github.com/open-telemetry/opentelemetry-collector/blob/e6f0e014c0ad280fc67557704a2015924caf76eb/config/confighttp/confighttp.go#L52

What problem are you experiencing with token passthrough?

File: exporter/signalfxexporter/exporter.go (hunk 3 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go:147]: As per the approach discussed, we will consider `maxIdleConns` value if provided, else take `maxConnections`’s value if provided and if not then we will assign a default value. So if we assign a default value in defaultConfig, and if the user has not provided the value of maxIdleConns then instead of checking the value of maxConnections it will directly take the default value of maxIdleConns from defaultconfig. So to check if the user has provided the value or not, we have not assigned the default value in defaultConfig and are assigning it in `getHTTPClient` function.

Yeah, I can define the default value as const but we still have to assign that const to a variable in this function as we have to assign an address to `maxIdleConns` as `maxIdleConns` is a pointer.

File: exporter/signalfxexporter/exporter.go (hunk 4 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go:164]: when access_token_passthrough is true, if we have SignalFx receiver and SignalFx exporter and an access token is provided to SignalFx receiver as header then that access token should be taken as an access token for the SignalFx exporter, irrespective of any user input of the exporter (as per current upstream implementation of the exporter). 
Now in SignalFx exporter config, if the user has provided access token in the header as “X-Sf-Token” and access_token_passthrough is true and if header(access token) is provided from SignalFx receiver then the access token passed from SignalFx receiver should be set as a header at the time of the client.Do(req) call, irrespective of any user input in the exporter.
Earlier we were calling the roundTrip function of net/http package where they were not overriding the request headers.  But when we use HTTPClientSettings, its [roundTrip function](https://github.com/open-telemetry/opentelemetry-collector/blob/51fcb65f02055638f223c6abd3d6d2d644c696db/config/confighttp/confighttp.go#L181) overrides the request headers and assigns the value to the user input access token. So at last the access token provided by the user will be passed, which is not expected behavior. Due to this, the following [unit test](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/38668ef013261b04e3de7b9d3a5bf0a100d7cbff/exporter/signalfxexporter/exporter_test.go#L302) was failing. 
So to resolve this we have set the headers of HTTPClientSettings to empty map. Still, if we want to retrieve our headers then they are available in headers config, so we have not dropped any of the headers.

Comment (rmfitzpatrick) [reply exporter/signalfxexporter/exporter.go:164]: Ok, I think I understand the issue and why this is required for the client creation.

File: exporter/signalfxexporter/config.go [hunk 5]
```diff
@@ -37,10 +38,10 @@ var _ config.Unmarshallable = (*Config)(nil)
...
 // Config defines configuration for SignalFx exporter.
 type Config struct {
-	config.ExporterSettings        `mapstructure:",squash"`
-	exporterhelper.TimeoutSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
-	exporterhelper.QueueSettings   `mapstructure:"sending_queue"`
-	exporterhelper.RetrySettings   `mapstructure:"retry_on_failure"`
+	config.ExporterSettings       `mapstructure:",squash"`
+	exporterhelper.QueueSettings  `mapstructure:"sending_queue"`
+	exporterhelper.RetrySettings  `mapstructure:"retry_on_failure"`
+	confighttp.HTTPClientSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/config.go:44]: I think we should delete the previous `Headers` field and rely on this one only since its squashed.  Operating on two maps throughout the exporter is unnecessary when we can just use the client settings one.

File: exporter/signalfxexporter/config_test.go [hunk 6]
```diff
@@ -58,17 +58,28 @@ func TestLoadConfig(t *testing.T) {
...
 	expectedCfg := Config{
 		ExporterSettings: config.NewExporterSettings(config.NewComponentIDWithName(typeStr, "allsettings")),
 		AccessToken:      "testToken",
 		Realm:            "us1",
-		MaxConnections:   70,
+		MaxConnections:   &maxConnections,
 		Headers: map[string]string{
 			"added-entry": "added value",
 			"dot.test":    "test",
 		},
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/config_test.go:73]: Since we're adding a new one in the client settings this field should be deleted/migrated everywhere.

File: exporter/signalfxexporter/exporter.go (hunk 4 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go]: If it looks good, could you please resolve this comment?

File: exporter/signalfxexporter/config.go (hunk 5 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/config.go:44]: Pushed the changes accordingly.

File: exporter/signalfxexporter/config_test.go (hunk 6 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/config_test.go]: Pushed the changes accordingly.

File: exporter/signalfxexporter/exporter.go (hunk 4 above)
Comment (rmfitzpatrick) [reply exporter/signalfxexporter/exporter.go]: I'm not able to resolve comments in this project

File: CHANGELOG.md [hunk 7]
```diff
@@ -40,6 +40,8 @@
 
 - `lokiexporter`: add complete log record to body (#6619)
 - `k8sclusterreceiver` add `container.image.tag` attribute (#6436)
+- `signalfxexporter`: add confighttp.HTTPClientSettings in config (#6803)
+
```
Comment (rmfitzpatrick) [inline CHANGELOG.md:44]:
```suggestion
```
nit

File: exporter/signalfxexporter/config.go [hunk 8]
```diff
@@ -160,7 +156,7 @@ func (cfg *Config) validateConfig() error {
 		return errors.New(`cannot have a negative "timeout"`)
 	}
 
-	if cfg.MaxConnections < 0 {
+	if cfg.MaxConnections != nil && *cfg.MaxConnections < 0 {
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/config.go:159]: Would probably be a good idea to log a warning that this field is deprecated if not nil and they should use the client settings fields.

File: exporter/signalfxexporter/exporter.go (hunk 4 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go]: okay, I do have access, I am resolving it.

File: CHANGELOG.md (hunk 7 above)
Comment (manang-splunk) [reply CHANGELOG.md]: Pushed the changes accordingly.

File: exporter/signalfxexporter/config.go (hunk 8 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/config.go]: Pushed the changes accordingly.

File: exporter/signalfxexporter/config.go [hunk 9]
```diff
@@ -160,8 +157,11 @@ func (cfg *Config) validateConfig() error {
 		return errors.New(`cannot have a negative "timeout"`)
 	}
 
-	if cfg.MaxConnections < 0 {
-		return errors.New(`cannot have a negative "max_connections"`)
+	if cfg.MaxConnections != nil {
+		logger.Warn(`"max_connections" is deprecated, use "max_idle_conns" or "max_idle_conns_per_host" instead`)
```
Comment (rmfitzpatrick) [inline exporter/signalfxexporter/config.go:161]: Realized we should probably have a check for the new values not being nil and log that only they will be used.

File: testbed/datasenders/signalfx.go [hunk 10]
```diff
@@ -65,7 +65,7 @@ func (sf *SFxMetricsDataSender) Start() error {
 	}
 
 	sf.Metrics = exporter
-	return nil
+	return exporter.Start(context.Background(), sf)
```
Comment (rmfitzpatrick) [inline testbed/datasenders/signalfx.go:68]: Why is this necessary?

File: exporter/signalfxexporter/config.go (hunk 9 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/config.go:161]: Pushed the changes accordingly

File: testbed/datasenders/signalfx.go (hunk 10 above)
Comment (manang-splunk) [reply testbed/datasenders/signalfx.go]: It was not necessary, therefore removed it.

**Review by rmfitzpatrick — APPROVED** [review]

File: exporter/signalfxexporter/exporter.go [hunk 11]
```diff
@@ -146,6 +145,33 @@ func newSignalFxExporter(
...
+			config.MaxIdleConnsPerHost = &defaultMaxConns
+		}
+	}
+
+	// Using another variable to avoid the local header changes from reflecting
+	// into actual config.
+	httpClientSettings := config.HTTPClientSettings
+	// To avoid overriding client headers to request header storing access_token
+	// when access_token_passthrough is true.
+	httpClientSettings.Headers = map[string]string{}
```
Comment (dmitryax) [inline exporter/signalfxexporter/exporter.go:170]: Aren't we completely discard `Headers` config here? I believe `pushMetricsDataForToken` takes care of setting token from access_token_passthrough

Comment (manang-splunk) [reply exporter/signalfxexporter/exporter.go:170]: We are not completely discarding the headers here. We have assigned `config.HTTPClientSettings` to different variable (`httpClientSettings`) and we are discarding headers of this variable. So the headers in the config will be there and are not discarded.
  You are right the `pushMetricsDataForToken` function is assigning [request headers](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bb538051fb03070d1683223dc10c381173fb9328/exporter/signalfxexporter/dpclient.go#L131) as per the access_token_passthrough, but after that when we are calling [client.Do(req)](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bb538051fb03070d1683223dc10c381173fb9328/exporter/signalfxexporter/dpclient.go#L140), it is calling the [roundTrip](https://github.com/open-telemetry/opentelemetry-collector/blob/4d0b6e4f891efca70a85a28ca6db74458b0dc935/config/confighttp/confighttp.go#L192) function of HTTPClient and in that the request headers are overridden by the client headers. So to avoid this we have assigned empty headers to client here.

File: exporter/signalfxexporter/factory_test.go [hunk 12]
```diff
@@ -134,10 +137,10 @@ func TestFactory_CreateMetricsExporterFails(t *testing.T) {
...
-				ExporterSettings: config.NewExporterSettings(config.NewComponentID(typeStr)),
-				AccessToken:      "testToken",
-				Realm:            "lab",
-				TimeoutSettings:  exporterhelper.TimeoutSettings{Timeout: -2 * time.Second},
+				ExporterSettings:   config.NewExporterSettings(config.NewComponentID(typeStr)),
+				AccessToken:        "testToken",
+				Realm:              "lab",
+				HTTPClientSettings: confighttp.HTTPClientSettings{Timeout: -2 * time.Second},
 			},
 			errorMessage: "failed to process \"signalfx\" config: cannot have a negative \"timeout\"",
```
Comment (MovieStoreGuy) [inline exporter/signalfxexporter/factory_test.go:145]: Can you also include a test for zero duration?

File: exporter/signalfxexporter/exporter.go (hunk 11 above)
Comment (dmitryax) [reply exporter/signalfxexporter/exporter.go:170]: Got it. Thanks for explanation

**Review by dmitryax — COMMENTED** [review]
Please address comment from @MovieStoreGuy, otherwise LGTM

File: exporter/signalfxexporter/factory_test.go (hunk 12 above)
Comment (manang-splunk) [reply exporter/signalfxexporter/factory_test.go:145]: This unit test is for testing failure, but as per current implementation for zero value we will not get an error and therefore we think we should not add it in this test. As the scope of this PR was to refactor the code to use HTTPConfigSettings, we haven’t changed the main implementation logic of this.
@dmitryax Please share your thoughts.

Comment (MovieStoreGuy) [reply exporter/signalfxexporter/factory_test.go:145]: A zero value is an interesting case since if directly setting the raw http client mean that the server must immediately return a result (which is physically possible) so it should be considered an error :)

**Comment by MovieStoreGuy** [comment]
@manang-splunk , are you still working on this PR?
